### PR TITLE
gtk4: init at 4.0.2

### PR DESCRIPTION
**Pull request by andir**
###### Motivation for this change
This is some initial work for getting GTK4 to compile as I wanted to get
a rough feeling how much effort it is.

https://gitlab.gnome.org/GNOME/gtk/-/blob/4.0.0/NEWS
https://blog.gtk.org/2020/12/16/gtk-4-0/

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
The pkg-config file is mangled preventing programs from building https://gitlab.gnome.org/GNOME/gtk/-/issues/3517.

File: pkgs/development/libraries/gtk/4.x.nix [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    for program in ''${demos[@]}; do
+      wrapProgram $dev/bin/$program \
+        --prefix XDG_DATA_DIRS : "$GSETTINGS_SCHEMAS_PATH:$out/share/gsettings-schemas/${pname}-${version}"
+    done
+  '';
+
+  passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = "gtk+";
+      attrPath = "gtk3";
```
Comment (jtojnar) [inline pkgs/development/libraries/gtk/4.x.nix:220]:
```suggestion
      packageName = "gtk";
      attrPath = "gtk4";
```

Comment (andir) [reply pkgs/development/libraries/gtk/4.x.nix:220]: Oh, maybe I should have made it more obvious: I do not plan to invest much/any time into this. I just recalled that GTK4 was a pity to build before the release and wanted to verify that again. Someone else - with an actual interest & motivation - should probably carry this forward in another PR that can be based on this one.

**Comment by tadfisher** [comment]
[4.0.2](https://gitlab.gnome.org/GNOME/gtk/-/tags/4.0.2) was shipped, so I think we can drop the patch.

Any other blockers to getting this merged?

**Comment by jtojnar** [comment]
The main burden is that we need to evaluate how does the GStreamer and GL support work – that is, does it crash without a wrapper.

File: pkgs/development/libraries/gtk/4.x.nix [hunk 2]
```diff
@@ -0,0 +1,229 @@
+{ stdenv
```
Comment (tadfisher) [inline pkgs/development/libraries/gtk/4.x.nix:1]:
```suggestion
{ lib
, stdenv
```

File: pkgs/development/libraries/gtk/4.x.nix [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+, cupsSupport ? stdenv.isLinux
+, withGtkDoc ? false && stdenv.isLinux
+, cups ? null
+, AppKit
+, Cocoa
+}:
+
+assert cupsSupport -> cups != null;
+
+with stdenv.lib;
```
Comment (tadfisher) [inline pkgs/development/libraries/gtk/4.x.nix:58]:
```suggestion
with lib;
```

**Comment by tadfisher** [comment]
@jtojnar Not sure if this is an issue with the vulkan package, but `nixpkgs#vulkan.dev` lists a non-existing `includedir` in `vulkan.pc`:

```
includedir=/nix/store/42qxa7x1v13c6bmbsl0cs94cicpygqma-vulkan-loader-1.2.162.0-dev/include
```

This is causing lots of directory-not-found warnings when building, but it does build. However, gstreamer applications fail to run because of a missing `playbin` element.

**Comment by tadfisher** [comment]
Need to add `graphene` to `propagatedBuildInputs` as well; at least gtk4-rs won't build without it.

**Comment by tadfisher** [comment]
Does it make sense to add `vulkan-loader` to `propagatedNativeBuildInputs` so downstream dependencies can locate vulkan via pkg-config?

**Comment by jtojnar** [comment]
I have no idea about GL stack but rule of thumb is: propagate when listed in `Requires` in a pkg-config file, do not propagate if listed in `Requires.private` or there are multiple libraries and only a small subset of them requires the dependency and the libraries are not commonly linked.

File: pkgs/development/libraries/gtk/4.x.nix [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+    for program in ''${demos[@]}; do
+      wrapProgram $dev/bin/$program \
+        --prefix XDG_DATA_DIRS : "$GSETTINGS_SCHEMAS_PATH:$out/share/gsettings-schemas/${pname}-${version}"
+    done
+  '';
+
+  passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = "gtk4";
```
Comment (jtojnar) [inline pkgs/development/libraries/gtk/4.x.nix:208]:
```suggestion
      packageName = "gtk";
```

File: pkgs/development/libraries/gtk/4.x.nix [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+, Cocoa
+}:
+
+assert cupsSupport -> cups != null;
+
+with lib;
+
+stdenv.mkDerivation rec {
+  pname = "gtk4";
+  version = "4.0.0";
```
Comment (jtojnar) [inline pkgs/development/libraries/gtk/4.x.nix:63]:
```suggestion
  version = "4.0.2";
```

File: pkgs/development/libraries/gtk/4.x.nix [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+  outputBin = "dev";
+
+  setupHooks = [
+    ./hooks/gtk4-clean-immodules-cache.sh
+    ./hooks/drop-icon-theme-cache.sh
+  ];
+
+  src = fetchurl {
+    url = "mirror://gnome/sources/gtk/${stdenv.lib.versions.majorMinor version}/gtk-${version}.tar.xz";
+    sha256 = "1ycv1cxh3s6d7hqjrkdi5kmpw1ml1i8fvxigbngds9za4yqzav6l";
```
Comment (jtojnar) [inline pkgs/development/libraries/gtk/4.x.nix:75]:
```suggestion
    sha256 = "115w3mzwm1xsi1q85qvwfm2yxpsjs2rcajgddzbnwhjicyn0frv2";
```

**Comment by jtojnar** [comment]
Let’s merge this for now. We will need to create `wrapGapps4Hook` but we can deal with that later.

**Comment by jtojnar** [comment]
Eh, turns out there has been a new release yesterday: https://gitlab.gnome.org/GNOME/gtk/-/blob/4.1.0/NEWS#L1-53

**Comment by andir** [comment]
Woohoo! Thanks for merging and working on this. I totally didn't except that to happen this quickly :-)

**Pull request merged**